### PR TITLE
This payment method is not available - Using a Billing address ( not available for payment method) differs from shipping address

### LINKED ISSUE #16723: This payment method is not available - Using a Billing address ( not available for payment method) differs from shipping address
**Describe the bug**
Issue reproduced by @Robin-Fischer-PS :) 
An error is displayed in the FO "This payment method is not available" after validation an order when using a Billing address ( not available for payment method) differs from shipping address

**To Reproduce**
Steps to reproduce the behavior: you must have two countries enabled (France & united states)
1. Go to BO => Payment => Preferences page => Country restrictions section
2. All payment modules are available on France but not available for the United States
image
3. Go to FO => add some products to the cart & process checkout
4. At the second step "Addresses", you must have two addresses
4.1. First address with France country  selected for the shipping address
4.2 Second address with United states selected for the billing address
5. Click on the "Billing address differs from shipping address" link
6. Continue checkout
7. Click on "Order with an obligation to pay" & see error: This payment method is not available.

**Screenshots**
https://drive.google.com/file/d/1AEZdUSbq1s0_VVFhzAspbp6yVwzwel0q/view

**Additional information**
PrestaShop version: 1.7.5.2 / 1.7.6.2
PHP version: 7.1

### PR DESCRIPTION
**Pull request by ghost**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop 
| Description?      | An error is displayed in the FO "This payment method is not available" after validation an order when using a Billing address ( not available for payment method) differs from shipping address
| Type?             | bug fix 
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #16723.
| How to test?      | Set up different payment methods to two countries and use the billing address different from the shipping address
| Possible impacts? | none.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24407)

### CODE REVIEW

File: classes/Hook.php [hunk 1]
```diff
@@ -1104,17 +1104,25 @@ private static function getAllHookRegistrations(Context $context, ?string $hookN
         } elseif ($frontend) {
             // For payment modules, we check that they are available in the contextual country
             if (Validate::isLoadedObject($context->country)) {
+                //Validate if the billing address is different from the delivery address in a valid cart.
```
Comment (Progi1984) [inline classes/Hook.php:1107]:
```suggestion
                // Validate if the billing address is different from the delivery address in a valid cart.
```

File: classes/Hook.php [hunk 2]
```diff
@@ -1104,17 +1104,25 @@ private static function getAllHookRegistrations(Context $context, ?string $hookN
...
+                $id_country = $context->country->id;
```
Comment (Progi1984) [inline classes/Hook.php:1108]:
```suggestion
                $id_country = (int) $context->country->id;
```

File: classes/Hook.php [hunk 3]
```diff
@@ -1104,17 +1104,25 @@ private static function getAllHookRegistrations(Context $context, ?string $hookN
...
+                if (Validate::isLoadedObject($context->cart) && $context->cart->id_address_delivery !== $context->cart->id_address_invoice) {
+                    $address_invoice = new Address($context->cart->id_address_invoice);
+                    if (Validate::isLoadedObject($address_invoice)) {
+                        $id_country = $address_invoice->id_country;
```
Comment (Progi1984) [inline classes/Hook.php:1112]:
```suggestion
                        $id_country = (int) $address_invoice->id_country;
```

File: classes/Hook.php [hunk 4]
```diff
@@ -1104,17 +1104,25 @@ private static function getAllHookRegistrations(Context $context, ?string $hookN
...
+                }
                 $sql->where(
                     '(
                         h.`name` IN ("displayPayment", "displayPaymentEU", "paymentOptions")
                         AND (
                             SELECT `id_country`
                             FROM `' . _DB_PREFIX_ . 'module_country` mc
                             WHERE mc.`id_module` = m.`id_module`
-                            AND `id_country` = ' . (int) $context->country->id . '
+                            AND `id_country` = ' . (int) $id_country . '
```
Comment (Progi1984) [inline classes/Hook.php:1122]:
```suggestion
                            AND `id_country` = ' . $id_country . '
```

File: classes/Hook.php [hunk 5]
```diff
@@ -1104,17 +1104,25 @@ private static function getAllHookRegistrations(Context $context, ?string $hookN
...
                             AND `id_shop` = ' . (int) $shop->id . '
                             LIMIT 1
-                        ) = ' . (int) $context->country->id . ')'
+                        ) = ' . (int) $id_country . ')'
```
Comment (Progi1984) [inline classes/Hook.php:1125]:
```suggestion
                        ) = ' . $id_country . ')'
```

**Review by matks — COMMENTED** [review]
Hello @presteamshop and thank you for your contribution ;) you got some suggestions from @Progi1984, can you have a look and apply them when relevant ?

**Review by ghost — COMMENTED** [review]
All perfect!

**Comment by PierreRambaud** [comment]
> All perfect!

I don't see any commits resolving what @Progi1984 asked :sweat_smile:

**Comment by PierreRambaud** [comment]
Hi @presteamshop ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Comment by presteamshop** [comment]
@PierreRambaud Sorry, I didn't know I should do something else, I'm not very related to Github.

**Pull request closed without merging**